dali/pipeline/operators/color/hsv.h

```diff
@@ -0,0 +1,284 @@
+                                              argument_t<GPUBackend> &arg) {
+  GetSingleOrRepeatedArg(spec, arg, arg_name);
+}
+
+
+template <typename Backend>
+struct WorkspaceInputType {
+  using type = TensorVector<CPUBackend>;
+};
+
+template <>
+struct WorkspaceInputType<GPUBackend> {
+  using type = TensorList<GPUBackend>;
+};
+
+template <typename Backend>
+using workspace_input_t = typename WorkspaceInputType<Backend>::type;
+
+
+}  // namespace hsv
+
+
+
+template <typename Backend>
+class Hsv : public Operator<Backend> {
+ public:
+  explicit Hsv(const OpSpec &spec) :
+          Operator<Backend>(spec),
+          output_type_(spec.GetArgument<DALIDataType>(hsv::kOutputType)) {
+    hsv::assign_argument_value<Backend>(spec, hsv::kHue, hue_);
+    hsv::assign_argument_value<Backend>(spec, hsv::kSaturation, saturation_);
+    hsv::assign_argument_value<Backend>(spec, hsv::kValue, value_);
+    if (std::is_same<Backend, GPUBackend>::value) {
+      kernel_manager_.Resize(1, 1);
+    } else {
+      kernel_manager_.Resize(num_threads_, batch_size_);
+    }
+  }
+
+  
+  ~Hsv() override = default;
+
+
+  DISABLE_COPY_MOVE_ASSIGN(Hsv);
+
+
+ protected:
+  bool CanInferOutputs() const override {
+    return true;
+  }
+
+
+  USE_OPERATOR_MEMBERS();
+  hsv::argument_t<Backend> hue_, saturation_, value_;
+  DALIDataType output_type_;
+  kernels::KernelManager kernel_manager_;
+};
+
+
+
+class HsvCpu : public Hsv<CPUBackend> {
+ public:
+  explicit HsvCpu(const OpSpec &spec) : Hsv(spec) {}
+
+  /*
+   * So that compiler wouldn't complain, that
+   * "overloaded virtual function `dali::Operator<dali::CPUBackend>::RunImpl`
+   * is only partially overridden in class `dali::HsvCpu`"
+   */
+  using Operator<CPUBackend>::RunImpl;
+
+  ~HsvCpu() override = default;
+
+
+  DISABLE_COPY_MOVE_ASSIGN(HsvCpu);
+
+
+ protected:
+  bool SetupImpl(std::vector<::dali::OutputDesc> &output_desc,
+                 const workspace_t<CPUBackend> &ws) override;
+
+
+  void RunImpl(Workspace<CPUBackend> &ws) override;
+
+
+ private:
+  mat3 transformation_matrix(float hue, float saturation, float value) const {
+    using namespace hsv;  // NOLINT
+    return Rgb2Yiq * compose_hue(hue) * compose_saturation(saturation) * compose_value(value) *
+           Yiq2Rgb;
+  }
+
+
+  template <typename Kernel, typename InputType>
+  kernels::TensorListShape<> CallSetup(const TensorVector<CPUBackend> &input, int instance_idx) {
+    kernels::KernelContext ctx;
+    kernels::TensorListShape<> sh = input.shape();
+    std::vector<kernels::TensorShape<>> shapes;
+    shapes.resize(sh.num_samples());
+    for (size_t i = 0; i < shapes.size(); i++) {
+      const auto tvin = view<const InputType, 3>(input[i]);
+      const auto tmat = transformation_matrix(hue_, saturation_, value_);
+      const auto reqs = kernel_manager_.Setup<Kernel>(instance_idx, ctx, tvin, tmat);
+      kernels::TensorListShape<> out_sh = reqs.output_shapes[0];
+      shapes[i] = out_sh.tensor_shape(0);
```

Comment:
Is this a mistake, or all images are assumed equal? But some of them are probably more equal than others.
```suggestion
      shapes[i] = out_sh.tensor_shape(i);
```

Reply:
Keeping a vector of TensorShapes is bad. Use a TensorListShape (as per Krzysztof's suggestion) and call 
```suggestion
      shapes.set_tensor_shape(i, reqs.output_shapes[0])
```

Reply:
Done